libkbfs/file_data.go

```diff
@@ -501,44 +527,291 @@ func (fd *fileData) createIndirectBlock(
+	df *dirtyFile, dver DataVer) (*FileBlock, []BlockPointer, error) {
+	// Find the lowest block that can accommodate a new right block.
+	lowestAncestorWithRoom := -1
+	for i := len(parentBlocks) - 1; i >= 0; i-- {
+		pb := parentBlocks[i]
+		if len(pb.pblock.IPtrs) < fd.bsplit.MaxPtrsPerBlock() {
+			lowestAncestorWithRoom = i
+			break
+		}
 	}
-	rblock := &FileBlock{}
 
-	newPtr := BlockPointer{
-		ID:      newRID,
-		KeyGen:  fd.kmd.LatestKeyGeneration(),
-		DataVer: DefaultNewBlockDataVersion(false),
-		Context: kbfsblock.Context{
-			Creator:  fd.uid,
-			RefNonce: kbfsblock.ZeroRefNonce,
-		},
+	var newTopBlock *FileBlock
+	var newDirtyPtrs []BlockPointer
+	if lowestAncestorWithRoom < 0 {
+		// Create a new level of indirection at the top.
+		var err error
+		newTopBlock, err = fd.createIndirectBlock(ctx, df, dver)
+		if err != nil {
+			return nil, nil, err
+		}
+
+		// The old top block needs to be cached under its new ID if it
+		// was indirect.
+		if len(parentBlocks) > 0 {
+			ptr := newTopBlock.IPtrs[0].BlockPointer
+			err = fd.cacher(ptr, parentBlocks[0].pblock)
+			if err != nil {
+				return nil, nil, err
+			}
+			newDirtyPtrs = append(newDirtyPtrs, ptr)
+		}
+
+		parentBlocks = append([]parentBlockAndChildIndex{{newTopBlock, 0}},
+			parentBlocks...)
+		lowestAncestorWithRoom = 0
 	}
 
-	pblock.IPtrs = append(pblock.IPtrs, IndirectFilePtr{
-		BlockInfo: BlockInfo{
-			BlockPointer: newPtr,
-			EncodedSize:  0,
-		},
-		Off: off,
-	})
+	fd.log.CDebugf(ctx, "Making new right block at off %d for file %v, "+
+		"lowestAncestor at level %d", off, fd.file.tailPointer(),
+		lowestAncestorWithRoom)
 
-	err = fd.cacher(newPtr, rblock)
-	if err != nil {
-		return err
+	// Make a new right block for every parent, starting with the
+	// lowest ancestor with room.
+	pblock := parentBlocks[lowestAncestorWithRoom].pblock
+	for i := lowestAncestorWithRoom; i < len(parentBlocks); i++ {
+		newRID, err := fd.crypto.MakeTemporaryBlockID()
+		if err != nil {
+			return nil, nil, err
+		}
+
+		newPtr := BlockPointer{
+			ID:      newRID,
+			KeyGen:  fd.kmd.LatestKeyGeneration(),
+			DataVer: dver,
+			Context: kbfsblock.Context{
+				Creator:  fd.uid,
+				RefNonce: kbfsblock.ZeroRefNonce,
+			},
+		}
+
+		fd.log.CDebugf(ctx, "New right block for file %d, level %d, ptr %v",
+			fd.file.tailPointer(), i, newPtr)
+
+		pblock.IPtrs = append(pblock.IPtrs, IndirectFilePtr{
+			BlockInfo: BlockInfo{
+				BlockPointer: newPtr,
+				EncodedSize:  0,
+			},
+			Off: off,
+		})
+
+		rblock := &FileBlock{}
+		if i != len(parentBlocks)-1 {
+			rblock.IsInd = true
+			pblock = rblock
+		}
+
+		err = fd.cacher(newPtr, rblock)
+		if err != nil {
+			return nil, nil, err
+		}
+
+		newDirtyPtrs = append(newDirtyPtrs, newPtr)
 	}
-	err = fd.cacher(ptr, pblock)
-	if err != nil {
-		return err
+
+	// All parents up to and including the lowest ancestor with room
+	// will have to change, so mark them as dirty.
+	ptr := fd.file.tailPointer()
```

Comment:
I'll make a `fileData.rootBlockPointer()` func that just returns `fd.file.tailPointer()`, for code clarity.